build_tools/setup_helpers/extension.py

```diff
@@ -40,150 +57,97 @@ def _get_ela(debug):
-        build_env.setdefault("CXX", "cl")
-    else:
-        extra_args = ["-DCMAKE_CXX_FLAGS=-fPIC " + _get_cxx11_abi()]
-    subprocess.run(
-        args=[
-            "cmake",
+    def build_extension(self, ext):
+        # Since two library files (libtorchaudio and _torchaudio) need to be
+        # recognized by setuptools, we instantiate `Extension` twice. (see `get_ext_modules`)
+        # This leads to the situation where this `build_extension` method is called twice.
+        # However, the following `cmake` command will build all of them at the same time,
+        # so, we do not need to perform `cmake` twice.
+        # Therefore we call `cmake` only for `torchaudio._torchaudio`.
+        if ext.name != "torchtext._torchtext":
+            return
+
+        extdir = os.path.abspath(os.path.dirname(self.get_ext_fullpath(ext.name)))
+
+        # required for auto-detection of auxiliary "native" libs
+        if not extdir.endswith(os.path.sep):
+            extdir += os.path.sep
+
+        cfg = "Debug" if self.debug else "Release"
+
+        cmake_args = [
+            f"-DCMAKE_BUILD_TYPE={cfg}",
+            f"-DCMAKE_PREFIX_PATH={torch.utils.cmake_prefix_path}",
+            f"-DCMAKE_INSTALL_PREFIX={extdir}",
+            "-DCMAKE_VERBOSE_MAKEFILE=ON",
+            f"-DPython_INCLUDE_DIR={distutils.sysconfig.get_python_inc()}",
+            "-DBUILD_TORCHTEXT_PYTHON_EXTENSION:BOOL=ON",
+            "-DRE2_BUILD_TESTING:BOOL=OFF",
+            "-DBUILD_TESTING:BOOL=OFF",
             "-DBUILD_SHARED_LIBS=OFF",
-            "-DRE2_BUILD_TESTING=OFF",
-            "-DCMAKE_EXPORT_COMPILE_COMMANDS=ON",
-            f"-DCMAKE_INSTALL_PREFIX={_TP_INSTALL_DIR}",
-            f"-DCMAKE_BUILD_TYPE={config}",
-            "-DCMAKE_CXX_VISIBILITY_PRESET=hidden",
             "-DCMAKE_POLICY_DEFAULT_CMP0063=NEW",
-        ]
-        + extra_args
-        + [".."],
-        cwd=str(build_dir),
-        check=True,
-        env=build_env,
-    )
-    print("*** Command list Thirdparty ***")
-    with open(build_dir / "compile_commands.json", "r") as fileobj:
-        print(fileobj.read())
-    print("running cmake --build", flush=True)
-    subprocess.run(
-        args=["cmake", "--build", ".", "--target", "install", "--config", config],
-        cwd=str(build_dir),
-        check=True,
-        env=build_env,
-    )
-
-
-def _build_sentence_piece(debug):
-    build_dir = _TP_BASE_DIR / "sentencepiece" / "build"
-    build_dir.mkdir(exist_ok=True)
-    build_env = os.environ.copy()
-    config = "Debug" if debug else "Release"
-    if platform.system() == "Windows":
-        extra_args = ["-GNinja"]
-        build_env.setdefault("CC", "cl")
-        build_env.setdefault("CXX", "cl")
-    else:
-        extra_args = []
-    subprocess.run(
-        args=[
-            "cmake",
+            "-DCMAKE_CXX_FLAGS=" + _get_cxx11_abi() + " " + _get_eca(cfg == "Debug"),
+            "-DCMAKE_EXE_LINKER_FLAGS=" + _get_ela(cfg == "Debug"),
```

Comment:
This affects flags of all the compilation.

In modern CMake it is recommended to approach this in more OOP-like fashion, using `target`s.

Once we verify this works, then we should move these dedicated location.

Reply:
Updated to use `target`s. Thanks for the feedback here and I'm glad we figured out that switching to this approach resolved the Windows errors.